### PR TITLE
main code modified and basic test cases has been added

### PR DESCRIPTION
**Pull request by chandanikumari**
# Changes

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/experimental/blob/master/CONTRIBUTING.md)
for more details._

### CODE REVIEW

**Comment by chandanikumari** [comment]
@wlynch please review.

File: tekdoc/main.go [hunk 1]
```diff
@@ -17,15 +17,100 @@ limitations under the License.
...
+	"os"
+	"text/template"
+)
 
-	_ "github.com/tektoncd/pipeline/pkg/pod"
+var (
+	filename = flag.String("f", "", "Name of the file to parse")
 )
 
+// TODO: take a file path, parse it as YAML -> v1alpha1.Task
```
Comment (wlynch) [inline tekdoc/main.go]: We can remove this TODO :D

File: tekdoc/main.go [hunk 2]
```diff
@@ -17,15 +17,100 @@ limitations under the License.
...
+
+func read(name string) (v1alpha1.Task, error) {
```
Comment (wlynch) [inline tekdoc/main.go]: Probably want this function name to be a little more descriptive so it's easier to understand exactly what it's doing. `readTask` perhaps?

File: tekdoc/main.go [hunk 3]
```diff
@@ -17,15 +17,100 @@ limitations under the License.
...
+		return task, fmt.Errorf("error reading file: %w", err)
+	}
+	if _, _, err := scheme.Codecs.UniversalDeserializer().Decode(dat, nil, &task); err != nil {
+		return task, fmt.Errorf("error decoding task: %w", err)
+	}
+
+	return task, nil
+}
+
+// TODO: use Go templates to print v1alpha1.Task as Markdown
```
Comment (wlynch) [inline tekdoc/main.go]: Remove TODO

File: tekdoc/main.go [hunk 4]
```diff
@@ -17,15 +17,100 @@ limitations under the License.
...
+	if err != nil {
+		log.Fatalln(err)
+	}
 
-	// TODO: take a file path, parse it as YAML -> v1alpha1.Task
-	// TODO: use Go templates to print v1alpha1.Task as Markdown
-	// TODO: tests!
 }
+
+// TODO: tests!
```
Comment (wlynch) [inline tekdoc/main.go]: Remove TODO

File: tekdoc/tekdoc_test.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+	"testing"
+)
+
+func TestTekdoc(t *testing.T) {
+	task, err := read("buildkit-daemonless.yaml")
+	if err != nil {
+		t.Fatal(err)
+	}
+	if task.Kind != "Task"{
+		t.Error("This is not a task file")
```
Comment (wlynch) [inline tekdoc/tekdoc_test.go]: We should include the `Kind` we got in the error message so it's easier to debug if/when this fails.

File: tekdoc/tekdoc_test.go [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+import (
+	"bytes"
+	"github.com/google/go-cmp/cmp"
+	"testing"
+)
+
+func TestTekdoc(t *testing.T) {
+	task, err := read("buildkit-daemonless.yaml")
+	if err != nil {
+		t.Fatal(err)
```
Comment (wlynch) [inline tekdoc/tekdoc_test.go]: Include a short error message about what failed (e.g. `t.Fatalf("failed to read task: %v", err)`)

File: tekdoc/tekdoc_test.go [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+	}
+	if task.Kind != "Task"{
+		t.Error("This is not a task file")
+	}
+
+	b := new(bytes.Buffer)
+	err = printTask(b,task)
+	if err != nil {
```
Comment (wlynch) [inline tekdoc/tekdoc_test.go]: if err := printTask(b,task); err != nil {

File: tekdoc/tekdoc_test.go [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+### Input:-
+- DOCKERFILE, The name of the Dockerfile
+- BUILDKIT_IMAGE, The name of the BuildKit image
+### Resources:-
+- source, git
+### Output:-
+- image, image
+`
+	if diff := cmp.Diff(v, b.String()); diff != ""{
+		t.Error(diff)
```
Comment (wlynch) [inline tekdoc/tekdoc_test.go]: nit: `t.Errorf("-want, +got: %s", diff)`

Makes it easier to read the output. :)

File: tekdoc/main.go [hunk 9]
```diff
@@ -17,15 +17,100 @@ limitations under the License.
...
+	return nil
+}
+
 func main() {
-	fmt.Println("Hello this is STILL TekDoc")
+	flag.Parse()
+
+	task, err := read(*filename)
+	if err != nil {
+		log.Fatalln(err)
```
Comment (wlynch) [inline tekdoc/main.go]: Include what failed in the error message here - log.Fatalln("failed to read Task:", err)

File: tekdoc/main.go [hunk 10]
```diff
@@ -17,15 +17,100 @@ limitations under the License.
...
+	}
+
+	err = printTask(os.Stdout, task)
+	if err != nil {
+		log.Fatalln(err)
```
Comment (wlynch) [inline tekdoc/main.go]: log.Fatalln("failed to render Task:", err)

File: tekdoc/main.go [hunk 11]
```diff
@@ -17,15 +17,100 @@ limitations under the License.
...
+	}
+
+	if task.Spec.Outputs != nil {
+
+		t := `### Output:-
+{{range .}}- {{.ResourceDeclaration.Name}}, {{.ResourceDeclaration.Type}}
+{{end}}`
+		tmpl := template.Must(template.New("test").Parse(t))
+		err := tmpl.Execute(w, task.Spec.Outputs.Resources)
+		if err != nil {
```
Comment (wlynch) [inline tekdoc/main.go]: if err := tmpl.Execute(w, task.Spec.Outputs.Resources); err != nil {

File: tekdoc/main.go [hunk 12]
```diff
@@ -17,15 +17,100 @@ limitations under the License.
...
+func printTask(w io.Writer, task v1alpha1.Task) error{
+	if task.Spec.Inputs != nil {
+
+		tmpl := template.Must(template.New("test").Parse(`# {{.Name}}
+## Install the Task
+kubectl apply -f https://raw.githubusercontent.com/tektoncd/catalog/master/{{.Name}}/{{.Name}}.yaml
+### Input:-
+`))
+		if err := tmpl.Execute(w, task); err != nil {
+			return fmt.Errorf("error executing the template: %v", err)
```
Comment (wlynch) [inline tekdoc/main.go]: Use %w instead of %v (here and below).

See https://blog.golang.org/go1.13-errors for more details about %w and its usage. :D

**Comment by bobcatfish** [comment]
Hey @chandanikumari !! Really excited to see this coming along :D

I want to leave a quick piece of feedback about the commit message "main code modified and basic test cases has been added". Although it's often not what people do, I find it really useful to put some time into writing fairly detailed commit messages about why the changes are being made and what they are.

There's a bit more info about this in our community repo: https://github.com/tektoncd/community/blob/master/standards.md#commit-messages

I also really like this article: https://chris.beams.io/posts/git-commit/

And lastly, shameless plug for my own talk where I talk about commit messages and some other stuff: https://www.youtube.com/watch?v=46f6lPmmy9s

File: tekdoc/main.go [hunk 13]
```diff
@@ -31,67 +32,83 @@ var (
...
+func main() {
+
+	flag.Parse()
+
+	task, err := readTask(*filename)
+	if err != nil {
+		log.Fatalln("failed to read Task:", err)
+	}
+
+	err = printTask(os.Stdout, task)
```
Comment (wlynch) [inline tekdoc/main.go]: if err := printTask(); err != nil {

File: tekdoc/tekdoc_test.go [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+	task, err := readTask("buildkit-daemonless.yaml")
+	if err != nil {
+		t.Fatalf("failed to read task: %v", err)
+	}
+	if task.Kind != "Task"{
+		t.Errorf("wanted Task file got %v", task.Kind)
+	}
+
+	b := new(bytes.Buffer)
+	if err = printTask(b,task); err != nil {
```
Comment (wlynch) [inline tekdoc/tekdoc_test.go]: if err := ...

File: tekdoc/main.go (hunk 13 above)
Comment (chandanikumari) [reply tekdoc/main.go]: done.

File: tekdoc/tekdoc_test.go (hunk 14 above)
Comment (chandanikumari) [reply tekdoc/tekdoc_test.go]: done.

**Pull request closed without merging**